### PR TITLE
CloudWatch: fix log explorer context

### LINKED ISSUE #41689: CloudWatch Logs: Non default regions does not seem to be sent for some queries
As seen here: https://github.com/grafana/grafana/pull/39290/files/e5e0ae1f9130208fdfc94b3d56021846386da153#diff-5195d59bff5abb37dceeb2ba37e2e0acb2f73f861fc83e320a72f6e7b2784f18R590 it seems like sometimes we don't send region from the query to the backend (like for autocomplete queries) which means we can get odd behavior if other then default region is used.

### PR DESCRIPTION
**Pull request by iwysiu**
**What this PR does / why we need it**:
Fixes the bug where `show context` in explore wouldn't work if the log query region wasn't the same as the default cloudwatch region.
Also fixes a second bug where we were sending a `0` endtime to the cloudwatch api if it wasn't set on the `GetLogEventsWithContext` query.

**Which issue(s) this PR fixes**:

Fixes #41689 

**Special notes for your reviewer**:
I don't love the test I ended up adding, but there wasn't really a good way to test this. If you prefer it, I can pass the `datasourceType` and `logQueries` from `Logs.tsx` down to `LogRowContextProvider.tsx` and test it there, but it seemed messier from a code perspective.

### CODE REVIEW

**Review by gabor — DISMISSED** [review]
i understand more data is needed in `datasource.gegLogRowContext` to fix the bug, but i'd recommend not adding cloudwatch-specific attributes to a `grafana-ui` structure. would something like this be enough? https://github.com/grafana/grafana/pull/52130
we can also have a chat about this and discuss possibilities, if you prefer so 👍

**Comment by iwysiu** [comment]
> i understand more data is needed in `datasource.gegLogRowContext` to fix the bug, but i'd recommend not adding cloudwatch-specific attributes to a `grafana-ui` structure. would something like this be enough? #52130 we can also have a chat about this and discuss possibilities, if you prefer so 👍

That would work! I'll pause on this until that goes in.

**Comment by iwysiu** [comment]
I altered it to use that `getLogRowContext` change instead. Also, I added a Go test.

**Review by iwysiu — COMMENTED** [review]
The changes to `DataSourceWithLogsContextSupport` have been added and it's ready to be reviewed!

**Review by gabor — COMMENTED** [review]
nice work (i cannot review the cloudwatch-specific code, i'm only adding this review so that my previous request-changes-review goes away)

EDIT: aaand, apparently this is not enough. please just dismiss my request-changes review and move forward with this 😄

File: public/app/plugins/datasource/cloudwatch/datasource.ts [hunk 1]
```diff
@@ -510,6 +511,7 @@ export class CloudWatchDatasource
     const requestParams: GetLogEventsRequest = {
       limit,
       startFromHead: direction !== 'BACKWARD',
+      region: query?.region,
```
Comment (sarahzinger) [inline public/app/plugins/datasource/cloudwatch/datasource.ts:514]: is there ever a situation where region is not defined? Will it still work? I could see a situation where you have to be like `query?.region || "__default"` or something like that, but I'm not 100% sure that's necessary, just a question to look into

**Review by sarahzinger — APPROVED** [review]
🚀

**Review by kevinwcyu — APPROVED** [review]

**Pull request merged**